sdk/eidolon_ai_sdk/cpu/conversational_apu.py

```diff
@@ -3,50 +3,79 @@
 from fastapi import HTTPException
 from opentelemetry import trace
 
-from eidolon_ai_sdk.cpu.agent_cpu import AgentCPU, AgentCPUSpec, Thread, CPUException
+from eidolon_ai_client.events import (
+    StreamEvent,
+    LLMToolCallRequestEvent,
+    ToolCallStartEvent,
+)
+from eidolon_ai_client.util.logger import logger
+from eidolon_ai_client.util.stream_collector import merge_streams
+from eidolon_ai_sdk.agent.doc_manager.document_processor import DocumentProcessor
+from eidolon_ai_sdk.agent.doc_manager.loaders.base_loader import FileInfo
+from eidolon_ai_sdk.agent.doc_manager.parsers.base_parser import DataBlob
+from eidolon_ai_sdk.agent_os import AgentOS
+from eidolon_ai_sdk.cpu.agent_cpu import APU, APUSpec, Thread, CPUException, APUCapabilities
 from eidolon_ai_sdk.cpu.agent_io import IOUnit, CPUMessageTypes
+from eidolon_ai_sdk.cpu.audio_unit import AudioUnit
 from eidolon_ai_sdk.cpu.call_context import CallContext
-from eidolon_ai_sdk.cpu.llm_message import AssistantMessage, ToolResponseMessage, LLMMessage
+from eidolon_ai_sdk.cpu.image_unit import ImageUnit
+from eidolon_ai_sdk.cpu.llm_message import AssistantMessage, ToolResponseMessage, LLMMessage, UserMessageFile, UserMessageAudio, UserMessageImage, \
+    UserMessageText, UserMessage
 from eidolon_ai_sdk.cpu.llm_unit import LLMUnit
 from eidolon_ai_sdk.cpu.logic_unit import LogicUnit, LLMToolWrapper
 from eidolon_ai_sdk.cpu.memory_unit import MemoryUnit
 from eidolon_ai_sdk.cpu.processing_unit import ProcessingUnitLocator, PU_T
-from eidolon_ai_client.events import (
-    StreamEvent,
-    LLMToolCallRequestEvent,
-    ToolCallStartEvent,
-)
 from eidolon_ai_sdk.system.reference_model import Reference, AnnotatedReference, Specable
-from eidolon_ai_client.util.logger import logger
 from eidolon_ai_sdk.util.stream_collector import StreamCollector, stream_manager, ManagedContextError
-from eidolon_ai_client.util.stream_collector import merge_streams
-
 
 tracer = trace.get_tracer("cpu")
 
 
-class ConversationalAgentCPUSpec(AgentCPUSpec):
+class ConversationalAPUSpec(APUSpec):
     io_unit: AnnotatedReference[IOUnit]
     memory_unit: AnnotatedReference[MemoryUnit]
     llm_unit: AnnotatedReference[LLMUnit]
     logic_units: List[Reference[LogicUnit]] = []
+    audio_unit: Reference[AudioUnit] = None
+    image_unit: Reference[ImageUnit] = None
     record_conversation: bool = True
     allow_tool_errors: bool = True
+    document_processor: AnnotatedReference[DocumentProcessor]
 
 
-class ConversationalAgentCPU(AgentCPU, Specable[ConversationalAgentCPUSpec], ProcessingUnitLocator):
+class ConversationalAPU(APU, Specable[ConversationalAPUSpec], ProcessingUnitLocator):
     io_unit: IOUnit
     memory_unit: MemoryUnit
     logic_units: List[LogicUnit]
+    llm_unit: LLMUnit
+    audio_unit: AudioUnit
```

Comment:
So what happens when you have a llm unit that supports images, and a separate image unit?

Presumably in this scenario we are specifying that we want the image pre-processed with the image processor? We should log probably log if this is the case for transparency.